adding documentation regarding CPU hotplug (#717)
kubevirt/user-guide: docs/operations/cpu_hotplug.md

# CPU Hotplug

> Starting kubevirt 1.0 its possible to configure the VM workload
> to allow for adding or removing the virtual CPUs while the VM is running.

Reviewer: ```suggestion
The CPU hotplug feature was introduced in KubeVirt v1.0, making it possible to configure the VM workload
to allow for adding or removing virtual CPUs while the VM is running.
```

### Abstract

> A **virtual CPU** is the CPU that is seen to the Guest VM OS. A VM owner can manage the amount of vCPUs from the VM spec template using the CPU topology fields (`spec.template.spec.domain.cpu`). The `cpu` object has the integers `cores,sockets,threads` so that the virtual CPU is calculated by the following formula: `cores * sockets * threads`. Before CPU hotplug was introduced the VM owner could change these integers in the VM template while the VM is running, and they were staged until the next boot cycle. With CPU hotplug it is possible to patch the `sockets` integer in the VM template and the change will take effect right away. The amount of vCPUs can be changed from 1 vCPU up to the limit of `cores * threads * maxSockets`. The `maxSockets` integer is pre-calculated by kubevirt but can also be managed by the VM owner or by the cluster administrator.

Reviewer: ```suggestion
A **virtual CPU** (vCPU) is the CPU that is seen to the Guest VM OS. A VM owner can manage the amount of vCPUs from the VM spec template using the CPU topology fields (`spec.template.spec.domain.cpu`). The `cpu` object has the integers `cores,sockets,threads` so that the virtual CPU is calculated by the following formula: `cores * sockets * threads`. 

Before CPU hotplug was introduced, the VM owner could change these integers in the VM template while the VM is running, and they were staged until the next boot cycle. With CPU hotplug, it is possible to patch the `sockets` integer in the VM template and the change will take effect right away. 

The amount of vCPUs can be changed from 1 vCPU up to the limit of `cores * threads * maxSockets`. The `maxSockets` integer is pre-calculated by KubeVirt but can also be managed by the VM owner or by the cluster administrator.
```

## Configuration

### Enable feature-gate
In order to enable CPU hotplug we need to add the `VMLiveUpdateFeatures` feature gate in Kubevirt CR:

````
apiVersion: kubevirt.io/v1
kind: KubeVirt
spec:
  configuration:
    developerConfiguration:
      featureGates:
        - VMLiveUpdateFeatures
````

> ### Enable the workload updater
> Workload updater must be configured with `LiveUpdate` strategy, as follows:

Reviewer: ```suggestion
### Configure the workload update strategy
The workload update strategy in the KubeVirt CR must be configured with `LiveUpdate`, as follows:
```
Suggesting this change because I couldn't find a reference to 'workload updater' in our docs, only to the 'VMI workload update strategy'.

Let's assume we have a running VM with the 4 vCPUs, which were configured with `sockets:4 cores:1 threads:1`
In the VMI status we can observe the current CPU topology the VM is running with:

```
apiVersion: kubevirt.io/v1
kind: VirtualMachineInstance
...
status:
  currentCPUTopology:
    cores: 1
    sockets: 4
    threads: 1
```

Now we want to hotplug another socket, by patching the VM object:

```
kubectl patch vm vm-cirros --type='json' \
-p='[{"op": "replace", "path": "/spec/template/spec/domain/cpu/sockets", "value": 5}]'
```
We can observe the CPU hotplug process in the VMI status:

```
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: null
    status: "True"
    type: LiveMigratable
  - lastProbeTime: null
    lastTransitionTime: null
    status: "True"
    type: HotVCPUChange
  currentCPUTopology:
    cores: 1
    sockets: 4
    threads: 1
```

Please note the condition `HotVCPUChange` that indicates the hotplug process is taking place.
Also you can notice the VirtualMachineInstanceMigration object that was creating for the VM in subject:

```
NAME                             PHASE     VMI
kubevirt-workload-update-kflnl   Running   vm-cirros

```

> Upon suceesful hotplug procedure, the `currentCPUTopology` will be updated with the new number of sockets, the migration
> would be marked as succesful.

Reviewer: ```suggestion
When the hotplug process has completed, the `currentCPUTopology` will be updated with the new number of sockets and the migration
is marked as successful.
```

# CPU Hotplug

The CPU hotplug feature was introduced in KubeVirt v1.0, making it possible to configure the VM workload
to allow for adding or removing virtual CPUs while the VM is running.

### Abstract
A **virtual CPU** (vCPU) is the CPU that is seen to the Guest VM OS. A VM owner can manage the amount of vCPUs from the VM spec template using the CPU topology fields (`spec.template.spec.domain.cpu`). The `cpu` object has the integers `cores,sockets,threads` so that the virtual CPU is calculated by the following formula: `cores * sockets * threads`. 

Before CPU hotplug was introduced, the VM owner could change these integers in the VM template while the VM is running, and they were staged until the next boot cycle. With CPU hotplug, it is possible to patch the `sockets` integer in the VM template and the change will take effect right away. 

> The amount of vCPUs can be changed from `cores * threads * 1` vCPU up to the limit of `cores * threads * maxSockets`. The `maxSockets` integer is pre-calculated by KubeVirt but can also be managed by the VM owner or by the cluster administrator.

Reviewer 2: ```suggestion
The amount of vCPUs can be changed from `cores * threads * sockets` vCPU up to the limit of `cores * threads * maxSockets`. The `maxSockets` integer is pre-calculated by KubeVirt but can also be managed by the VM owner or by the cluster administrator.
```

## Configuration

### Enable feature-gate
In order to enable CPU hotplug we need to add the `VMLiveUpdateFeatures` feature gate in Kubevirt CR:

````
apiVersion: kubevirt.io/v1
kind: KubeVirt
spec:
  configuration:
    developerConfiguration:
      featureGates:
        - VMLiveUpdateFeatures
````

### Configure the workload update strategy
The workload update strategy in the KubeVirt CR must be configured with `LiveUpdate`, as follows:

```
apiVersion: kubevirt.io/v1
kind: KubeVirt
spec:
  workloadUpdateStrategy:
    workloadUpdateMethods:
    - LiveMigrate
```

### Enable in VM spec
The VM object should look like this:
````
apiVersion: kubevirt.io/v1
kind: VirtualMachine
spec:
  liveUpdateFeatures:
    cpu: {}
````

> **NOTE:** In order for these changes to take effect the VM needs to be rebooted.

Reviewer 2: In that case, maxSockets will default to 2*sockets IIRC, right @enp0s3 ?
Probably worth adding here.

Author: Done

### [OPTIONAL] Set maximum sockets

> You can explicitly set the maximum amount of sockets in two ways - VM level or Cluster level

Reviewer 2: Which one takes precedence if both are set? (VM I assume)

```
apiVersion: kubevirt.io/v1
kind: VirtualMachineInstance
...
status:
  currentCPUTopology:
    cores: 1
    sockets: 4
    threads: 1
```

Now we want to hotplug another socket, by patching the VM object:

```
kubectl patch vm vm-cirros --type='json' \
-p='[{"op": "replace", "path": "/spec/template/spec/domain/cpu/sockets", "value": 5}]'
```
We can observe the CPU hotplug process in the VMI status:

```
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: null
    status: "True"
    type: LiveMigratable
  - lastProbeTime: null
    lastTransitionTime: null
    status: "True"
    type: HotVCPUChange
  currentCPUTopology:
    cores: 1
    sockets: 4
    threads: 1
```

Please note the condition `HotVCPUChange` that indicates the hotplug process is taking place.
Also you can notice the VirtualMachineInstanceMigration object that was creating for the VM in subject:

```
NAME                             PHASE     VMI
kubevirt-workload-update-kflnl   Running   vm-cirros

```
When the hotplug process has completed, the `currentCPUTopology` will be updated with the new number of sockets and the migration

> is marked as successful.

Reviewer 2: Can we see the updated spec/status?

Author: Added

# CPU Hotplug

The CPU hotplug feature was introduced in KubeVirt v1.0, making it possible to configure the VM workload
to allow for adding or removing virtual CPUs while the VM is running.

### Abstract
A **virtual CPU** (vCPU) is the CPU that is seen to the Guest VM OS. A VM owner can manage the amount of vCPUs from the VM spec template using the CPU topology fields (`spec.template.spec.domain.cpu`). The `cpu` object has the integers `cores,sockets,threads` so that the virtual CPU is calculated by the following formula: `cores * sockets * threads`. 

Before CPU hotplug was introduced, the VM owner could change these integers in the VM template while the VM is running, and they were staged until the next boot cycle. With CPU hotplug, it is possible to patch the `sockets` integer in the VM template and the change will take effect right away. 

Per each new socket that is hot-plugged, the amount of new vCPUs that would be seen by the guest is `cores * threads`, since the overall calculation of vCPUs is `cores * sockets * threads`. 

## Configuration

### Enable feature-gate
In order to enable CPU hotplug we need to add the `VMLiveUpdateFeatures` feature gate in Kubevirt CR:

> ````

Author: Done

apiVersion: kubevirt.io/v1
kind: KubeVirt
spec:
  configuration:
    developerConfiguration:
      featureGates:
        - VMLiveUpdateFeatures
````

### Configure the workload update strategy

> The workload update strategy in the KubeVirt CR must be configured with `LiveUpdate`, as follows:

Author: Done

```
apiVersion: kubevirt.io/v1
kind: KubeVirt
spec:
  workloadUpdateStrategy:
    workloadUpdateMethods:
    - LiveMigrate
```

### Enable in VM spec
The VM object should look like this:
````
apiVersion: kubevirt.io/v1
kind: VirtualMachine
spec:
  liveUpdateFeatures:
    cpu: {}
````
Specifying the `cpu.maxSockets` value is optional. If you leave it unset, it will default to 4 times the `sockets` value.

**NOTE:** In order for these changes to take effect the VM needs to be rebooted.

### [OPTIONAL] Set maximum sockets
You can explicitly set the maximum amount of sockets in two ways - VM level or Cluster level

> <table>

Reviewer 2: Nit: The table is a great idea, but it's pretty small on the preview. Maybe a `width:100%` or similar can be added to fix that.

Author: Done

Reviewer 2: /approve
Note: the table is still small, but that's fine. I'm not sure what the `<img width="100%">` lines do, but they should probably end with `/>` to be proper xml.

spec:
  configuration:
    liveUpdateConfiguration:
      maxCpuSockets: 8
````

</td>
</tr>
</table>

The VM-level configuration will take precendence over the cluster-wide configuration.

## Hotplug process
Let's assume we have a running VM with the 4 vCPUs, which were configured with `sockets:4 cores:1 threads:1`
In the VMI status we can observe the current CPU topology the VM is running with:

```
apiVersion: kubevirt.io/v1
kind: VirtualMachineInstance
...
status:
  currentCPUTopology:
    cores: 1
    sockets: 4
    threads: 1
```

Now we want to hotplug another socket, by patching the VM object:

```
kubectl patch vm vm-cirros --type='json' \
-p='[{"op": "replace", "path": "/spec/template/spec/domain/cpu/sockets", "value": 5}]'
```
We can observe the CPU hotplug process in the VMI status:

```
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: null
    status: "True"
    type: LiveMigratable
  - lastProbeTime: null
    lastTransitionTime: null
    status: "True"
    type: HotVCPUChange
  currentCPUTopology:
    cores: 1
    sockets: 4
    threads: 1
```

Please note the condition `HotVCPUChange` that indicates the hotplug process is taking place.

> Also you can notice the VirtualMachineInstanceMigration object that was creating for the VM in subject:

Author: Done

```
NAME                             PHASE     VMI
kubevirt-workload-update-kflnl   Running   vm-cirros

> 

Reviewer 2: Nit: empty line

Author: Done

A **virtual CPU** (vCPU) is the CPU that is seen to the Guest VM OS. A VM owner can manage the amount of vCPUs from the VM spec template using the CPU topology fields (`spec.template.spec.domain.cpu`). The `cpu` object has the integers `cores,sockets,threads` so that the virtual CPU is calculated by the following formula: `cores * sockets * threads`. 

Before CPU hotplug was introduced, the VM owner could change these integers in the VM template while the VM is running, and they were staged until the next boot cycle. With CPU hotplug, it is possible to patch the `sockets` integer in the VM template and the change will take effect right away. 

Per each new socket that is hot-plugged, the amount of new vCPUs that would be seen by the guest is `cores * threads`, since the overall calculation of vCPUs is `cores * sockets * threads`. 

## Configuration

### Enable feature-gate
In order to enable CPU hotplug we need to add the `VMLiveUpdateFeatures` feature gate in Kubevirt CR:

```yaml
apiVersion: kubevirt.io/v1
kind: KubeVirt
spec:
  configuration:
    developerConfiguration:
      featureGates:
        - VMLiveUpdateFeatures
```

### Configure the workload update strategy

> The workload update strategy in the KubeVirt CR must be configured with `LiveMigrate`, as follows:

Reviewer 3: Maybe it is worth it to explain from the outset the reason of this: hotplug requires live migration

Reviewer 3: Or just move the limitations section at the beginning?

Author: @fossedihelm Thanks! I've updated the section with more details

## Configuration

### Enable feature-gate
In order to enable CPU hotplug we need to add the `VMLiveUpdateFeatures` feature gate in Kubevirt CR:

```yaml
apiVersion: kubevirt.io/v1
kind: KubeVirt
spec:
  configuration:
    developerConfiguration:
      featureGates:
        - VMLiveUpdateFeatures
```

### Configure the workload update strategy
Current implementation of the hotplug process requires the VM to live-migrate.
The migration will be triggered automatically by the workload updater. The workload update strategy in the KubeVirt CR must be configured with `LiveMigrate`, as follows:

```yaml
apiVersion: kubevirt.io/v1
kind: KubeVirt
spec:
  workloadUpdateStrategy:
    workloadUpdateMethods:
    - LiveMigrate
```

### Enable in VM spec
The VM object should look like this:
```yaml
apiVersion: kubevirt.io/v1
kind: VirtualMachine
spec:
  liveUpdateFeatures:
    cpu: {}
```
Specifying the `cpu.maxSockets` value is optional. If you leave it unset, it will default to 4 times the `sockets` value.

> **NOTE:** In order for these changes to take effect the VM needs to be rebooted.

Reviewer 3: Out of this spefic PR but at some point in the future, we should decide whether to write notes as
`**NOTE:**`
or
`*NOTE:*`
or
`NOTE:`
or
`> **Note**`
or
`> *Note:*`
or
`> Note:`

@aburdenthehand FYI
